File: Eino testing/backend/advisor.go
package main

import (
	"context"
	"fmt"
	"strings"

	"github.com/cloudwego/eino-ext/components/model/openai"
	"github.com/cloudwego/eino/schema"
)

type Advisor struct {
	chatModel *openai.ChatModel
}

func NewAdvisor() (*Advisor, error) {
	ctx := context.Background()

	chatModel, err := openai.NewChatModel(ctx, &openai.ChatModelConfig{
		APIKey: getOpenAIKey(),
		Model:  AdvisorModel,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create chat model: %w", err)
	}

	return &Advisor{
		chatModel: chatModel,
	}, nil
}

func (a *Advisor) GenerateAdvice(building *Building, query string, retrievedDocs []RetrievedDoc) (string, []Citation, error) {
	ctx := context.Background()

	// Build query if not provided
	if query == "" {
		query = fmt.Sprintf(
			"Gi råd om energieffektivisering for en %d år gammel bygning på %.1f m² av type %s som bruker %.0f kWh/år, mens forventet bruk er %.0f kWh/år.",
			building.Age,
			building.SizeM2,
			building.ConstructionType,
			building.CurrentEnergyKwh,
			building.ExpectedEnergyKwh,
		)
	}

	// Format context from retrieved documents
	contextParts := []string{}
	for _, doc := range retrievedDocs {
		contextParts = append(contextParts, fmt.Sprintf("%s\n%s", doc.Citation, doc.Source))
	}
	context := strings.Join(contextParts, "\n\n---\n\n")

	// Format building data
	buildingText := formatBuildingForPrompt(building)

	// Build user prompt
	userPrompt := fmt.Sprintf(`%s

Relevante dokumenter:
%s

Basert på bygningsdataene og dokumentasjonen over, gi konkrete råd om energieffektivisering.
Husk å:
- Referere eksplisitt til bygningsdataene
- Sitere kildene du bruker
- Bruke rådgivende tone (foreslå, ikke instruer)
- Skille mellom fakta og antagelser
- Anerkjenne usikkerheter og begrensninger`, buildingText, context)

	// Create messages
	messages := []*schema.Message{
		{
			Role:    schema.System,
			Content: AdvisorSystemPrompt,
		},
		{
			Role:    schema.User,
			Content: userPrompt,
		},
	}

	// Generate advice
	result, err := a.chatModel.Generate(ctx, messages)
	if err != nil {
		return "", nil, fmt.Errorf("failed to generate advice: %w", err)
	}

	// Extract citations
	citations := extractCitations(result.Content, retrievedDocs)

	return result.Content, citations, nil
}

func formatBuildingForPrompt(building *Building) string {
	excessPercent := ((building.CurrentEnergyKwh - building.ExpectedEnergyKwh) / building.ExpectedEnergyKwh) * 100

	return fmt.Sprintf(`Bygningsdata:
- Navn: %s
- Alder: %d år
- Størrelse: %.1f m²
- Konstruksjonstype: %s
- Bygningstype: %s
- Nåværende energibruk: %.0f kWh/år
- Forventet energibruk: %.0f kWh/år
- Overskridelse: %.1f%%

Detaljer:
- Isolasjon: %s
- Vinduer: %s
- Oppvarmingssystem: %s
- Ventilasjon: %s
- Taktype: %s
- Orientering: %s`,
		building.Name,
		building.Age,
		building.SizeM2,
		building.ConstructionType,
		building.BuildingType,
		building.CurrentEnergyKwh,
		building.ExpectedEnergyKwh,
		excessPercent,
		building.Details["insulation"],
		building.Details["windows"],
		building.Details["heating_system"],
		building.Details["ventilation"],
		building.Details["roof_type"],
		building.Details["orientation"],
	)
}

func extractCitations(adviceText string, retrievedDocs []RetrievedDoc) []Citation {
	citations := []Citation{}

	for _, doc := range retrievedDocs {
		if strings.Contains(adviceText, doc.Citation) {
			citations = append(citations, Citation{
				Citation: doc.Citation,
				Source:   doc.Source,
				Page:     doc.Page,
			})
		}
	}

	return citations
}

File: Eino testing/advisor.py
"""
RAG-based advisor for generating energy advice with citations.
"""

import os
from typing import Dict, Any, Optional
from dotenv import load_dotenv

from config import ADVISOR_SYSTEM_PROMPT, ADVISOR_MODEL
from retriever import hybrid_retrieve, format_context_for_prompt
from building_data import format_building_for_prompt
from eino_client import create_eino_client

load_dotenv()


class EnergyAdvisor:
    """RAG-based energy advisor."""
    
    def __init__(self, api_key: Optional[str] = None, base_url: Optional[str] = None):
        """Initialize advisor with Eino client."""
        self.client = create_eino_client(api_key, base_url)
        self.model = ADVISOR_MODEL
    
    def generate_advice(self, building_data: Dict[str, Any], 
                      query: Optional[str] = None) -> Dict[str, Any]:
        """
        Generate energy advice for a building using RAG.
        
        Args:
            building_data: Building data dictionary
            query: Optional specific query, otherwise generates general advice
        
        Returns:
            Dictionary with advice text, citations, and metadata
        """
        # Build query if not provided
        if not query:
            query = f"""
            Gi råd om energieffektivisering for en {building_data.get('age', '')} år gammel bygning 
            på {building_data.get('size_m2', '')} m² av type {building_data.get('construction_type', '')} 
            som bruker {building_data.get('current_energy_kwh', '')} kWh/år, 
            mens forventet bruk er {building_data.get('expected_energy_kwh', '')} kWh/år.
            """
        
        # Retrieve relevant documents
        retrieved_docs = hybrid_retrieve(query, building_data)
        
        if not retrieved_docs:
            return {
                "advice": "Beklager, jeg fant ikke relevant dokumentasjon for å gi råd.",
                "citations": [],
                "retrieved_docs": [],
                "error": "Ingen dokumenter funnet"
            }
        
        # Format context
        context = format_context_for_prompt(retrieved_docs)
        
        # Format building data
        building_text = format_building_for_prompt(
            type('Building', (), building_data)()  # Simple object-like access
        )
        
        # Build user prompt
        user_prompt = f"""{building_text}

Relevante dokumenter:
{context}

Basert på bygningsdataene og dokumentasjonen over, gi konkrete råd om energieffektivisering.
Husk å:
- Referere eksplisitt til bygningsdataene
- Sitere kildene du bruker
- Bruke rådgivende tone (foreslå, ikke instruer)
- Skille mellom fakta og antagelser
- Anerkjenne usikkerheter og begrensninger"""
        
        # Generate advice with LLM
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": ADVISOR_SYSTEM_PROMPT},
                    {"role": "user", "content": user_prompt}
                ],
                temperature=0.7,
                max_tokens=1500
            )
            
            advice_text = response["choices"][0]["message"]["content"]
            
            # Extract citations from advice text
            citations = self._extract_citations(advice_text, retrieved_docs)
            
            # Handle usage - can be dict or object
            tokens_used = None
            if "usage" in response:
                if isinstance(response["usage"], dict):
                    tokens_used = response["usage"].get("total_tokens")
                else:
                    tokens_used = response["usage"].total_tokens if response["usage"] else None
            
            return {
                "advice": advice_text,
                "citations": citations,
                "retrieved_docs": [
                    {
                        "source": doc["source"],
                        "page": doc["page"],
                        "citation": doc["citation"]
                    }
                    for doc in retrieved_docs
                ],
                "metadata": {
                    "model": self.model,
                    "tokens_used": tokens_used,
                    "num_sources": len(retrieved_docs)
                }
            }
            
        except Exception as e:
            return {
                "advice": f"Feil ved generering av råd: {str(e)}",
                "citations": [],
                "retrieved_docs": [],
                "error": str(e)
            }
    
    def _extract_citations(self, advice_text: str, 
                          retrieved_docs: list) -> list:
        """Extract citations mentioned in advice text."""
        citations_found = []
        
        for doc in retrieved_docs:
            citation = doc.get("citation", "")
            if citation in advice_text:
                citations_found.append({
                    "citation": citation,
                    "source": doc.get("source", ""),
                    "page": doc.get("page", "")
                })
        
        return citations_found


def create_advisor() -> EnergyAdvisor:
    """Factory function to create advisor."""
    return EnergyAdvisor()


if __name__ == "__main__":
    # Test advisor
    from building_data import generate_buildings, get_building_dict
    
    buildings = generate_buildings(1)
    building_dict = get_building_dict(buildings[0])
    
    advisor = create_advisor()
    result = advisor.generate_advice(building_dict)
    
    print("RÅD:")
    print(result["advice"])
    print("\nSITATER:")
    for cit in result["citations"]:
        print(f"  - {cit['citation']}")
